[FIX] Specify marker file names for KIT data (MEG)

monkeyman192 opened the pull request:
closes #45

jasmainak commented:
Looks reasonable to me

chrisfilo commented:
Thank you for opening the PR.

I got one concern: you are using the `acq` filename keyword which is free form everywhere else but in this particular case you only two values are allowed. This makes the spec inconsistent. I would suggest proposing another keyword.

monkeyman192 commented:
I just did this because it's what @robertoostenveld and @teonbrooks recommended.
I agree however that there is the potential of conflict by using `acq`, hence why I initially suggested to just suffix `marker` with `pre` or `post`.
Essentially making the line
```
sub-control01_ses-001_task-rest_run-01_markers[-pre,-post].<mrk,sqd>
```

chrisfilo commented:
I would prefer new tag (like `marker`) instead of reusing `acq` for reasons stated above, but we need more MEG people to join the conversation.

teonbrooks commented:
since `_markers` is only specified for KIT in the spec, i think we could generalize it to a keyword as @chrisfilo suggested. I think having `-pre` and `-post` as its possible values is fine.

jasmainak commented:
What would be the suffix if the key-value pair is `marker-pre` and `marker-post` ? Shouldn't all files in BIDS follow the format:

``<key1>-<value1>_<key2>-<value2>_..._<suffix>.<extension>``

?

chrisfilo commented:
This is a good question. There should be a prefix.

monkeyman192 commented:
Could just reuse `meg`?
although it would then not make sense for when you don't specify `pre` or `post`...

jasmainak commented:
I think the most backwards compatible solution would be to leave the suffix as it is (`_markers.mrk`) and instead come up with a new key-value pair to indicate weather it's recorded before or after a trial.

monkeyman192 commented:
so I was looking through the specification for something else and noticed that the `recording` kvp is completely unused for MEG as far as I can tell, so maybe we could use this?
More specifically the name of the file would become:
`sub-<label>[_ses-<label>][_task-<label>][_run-<label>][_recording-<pre,post>]_markers.<mrk,sqd>`
Also, I think limiting this to two files is fine because you can't have more than two anyway I think.

chrisfilo commented:
I think `recording` is already used in physiological data: https://github.com/bids-standard/bids-specification/blob/master/src/04-modality-specific-files/04-physiological-and-other-continous-recordings.md#physiological-and-other-continuous-recordings

monkeyman192 commented:
would it matter if were used for a different context for a pretty niche case?
I don't think it would cause any conflict as `recording` isn't used in any other context for MEG data

chrisfilo commented:
I would prefer to avoid such internal inconsistency by picking a different keyword.

jasmainak commented:
@chrisfilo what would you suggest that maintains internal consistency? I suppose trying to use any of the existing keys would lead to consistency issues, so adding a new key-value pair seems to be the only solution.

chrisfilo commented:
Yes. I think adding a new keyword instead of redefining an old one would improve the consistency and make the standard more intuitive for people working with multiple modalities.

monkeyman192 commented:
I completely see the justification for wanting to introduce another kvp to avoid inconsistencies, however it just seems silly to add another for just one single case.
I am not *against* adding another kvp, I just think we keep adding synonyms for recording (`run`, `acquisition`), sometimes it may be *more* confusing having more terms. When someone looks at the specification they may wonder why there are so many terms that all mean essentially the same thing.

Finally do you have a suggestion for an appropriate key? I can't really think of one that would be good. If we don't care about having the kvp used elsewhere then we could just use `mark-<pre,post>_marker` even if it sort of doubles up the `marker` tag.

robertoostenveld commented:
I propose to consider this also in relation to EEG and especially iEEG specifications. For the iEEG it is  not uncommon to have pre and post-operative data. For EEG it would be possible (but not common) to have pre- and post-measurement electrode digitizations.

robertoostenveld commented:
besides the two `pre` and `post`, I think that more than two files with auxiliary data (i.e. data that enriches the core _meg/_eeg/_ieeg dataset) would also be possible. The labels `pre` and `post` are just two possible values. E.g. consider the case that multiple photos are taken of EEG electrodes or MEG markers, like with [photogrammetry](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5432580/). The series of photos would be from different angles, but not from different moments per see.

monkeyman192 commented:
maybe having `meas` (for `measurement`) as the `key` could work?
For the marker measurements we could say that the values have to be one of `pre` or `post`.
For other cases such as the photogrammetry the values could be un-restricted.

monkeyman192 commented:
Any further thoughts on this? It would be quite nice to get this sorted as it blocks a few things that I am working on and the sooner it is sorted the less impact it has on all the data converted to BIDS format by where I work.

robertoostenveld commented:
> What would be the suffix if the key-value pair is `marker-pre` and `marker-post` ? Shouldn't all filenames in BIDS follow the format:

The suffix is already specified as headshape, i.e. as
`sub-<label>[_ses-<label>][_acq-<label>]_headshape.<manufacturer_specific_extension>`

In the thread above `marker` is suggested as suffix. I don't see the need for that, and I don't see it anywhere in the current specification.

Re: restricted versus free form values: I think that just `pre` and `post` applies often, but will not enough in practice. Therefore I think that the value should free form and hence `acq` as the key would work.  I have a general preference of not introducing too many keys, especially not if they conceptually have the same meaning.

jasmainak commented:
I agree with the views shared above by @robertoostenveld and @sappelhoff . We should reuse the `acq` key

monkeyman192 commented:
ok, I have updated the PR to fix the merge conflict and update the text to reflect the above discussion

chrisfilo commented:
From a perspective of someone writing software that will ingest such datasets the current wording seems to liberal. For example, it would be completely ok for someone to use `acq-pre` for acquisition after  MEG or to encode something completely different. To make this semantic usable for data consumers I think you should go to semi-controlled vocabulary where `post` and `pre` are reserved keywords that have a singular meaning.

robertoostenveld commented:
why do you consider it required to be (semi) controlled here, and not for the initial use of the `_acq` label. For reference, this is what the specification contains: 

> The OPTIONAL acq-<label> key/value pair corresponds to a custom label the user MAY use to distinguish a different set of parameters used for acquiring the same modality. For example this should be used when a study includes two T1w images - one full brain low resolution and and one restricted field of view but high resolution. In such case two files could have the following names: sub-01_acq-highres_T1w.nii.gz and sub-01_acq-lowres_T1w.nii.gz, however the user is free to choose any other label than highres and lowres as long as they are consistent across subjects and sessions. In case different sequences are used to record the same modality (e.g. RARE and FLASH for T1w) this field can also be used to make that distinction. At what level of detail to make the distinction (e.g. just between RARE and FLASH, or between RARE, FLASH, and FLASHsubsampled) remains at the discretion of the researcher.

In the use for MEG headshape recordings, I would also say that " the user is free to choose any other label than pre and post as long as they are consistent across subjects and sessions".

chrisfilo commented:
Hi @robertoostenveld,

To be honest I would prefer is a different filename keyword was used for this purpose (see my comments above) for the same reason you mentioned - consistency with the usage of `acq` in other contexts. I suggested this solution as a compromise seeing the opposition towards adding new keywords.

As for the suggestion to "let users choose any label they want as long as it is consistent". I think this is what users would do with the current spec. The addition of a more strict controlled vocabulary enables programmatic uses. A piece of software could in such situation automatically find which files recording was made before MEG without having to ask the user for corresponding `acq` label.

monkeyman192 commented:
@chrisfilo what keyword would you suggest be used instead of `acq`?
I agree it is good to have consistency, but it seems like most others are pretty happy with using `acq`.
Also, @robertoostenveld I somewhat disagree with
> I think that just pre and post applies often, but will not enough in practice.

What other situations do you envisage?
In terms of the recording process where I work there will only ever be at most 2 markers associated with any one .con file, one before and one afterwards.

yarikoptic commented:
Although I am typically the one demanding semantically clear field, here `acq` seems to fit the bill, especially if we keep in mind that we are taking about  `_markers`.
If someone needs programmatic treatment of those files, I think it would be ok to demand specific pair of values, defaulting to pre and post. Specification could note that those are suggested values to describe pre/post data collections

yarikoptic commented on src/99-appendices/06-meg-file-formats.md:
````diff
@@ -180,10 +180,17 @@ Where:
 ```Text
 sub-control01_ses-001_task-rest_run-01_meg/
-    sub-control01_ses-001_task-rest_run-01_markers.<mrk,sqd>
+    sub-control01_ses-001_task-rest_run-01[_acq-<acq-label>]_markers.<mrk,sqd>
````
Isn't there also a table to adjust in the supplementals for having acq allowed here? Sorry, hard to lookup on the phone atm

monkeyman192 replied:
yes! Good call! I forgot about that table.

monkeyman192 replied:
Actually, I think looking at the table it is already included. `acq` is already an optional parameter for `meg`, and there is probably no point in added `marker` to the top column title since it is only for a single manufacturer...

chrisfilo replied:
it might be worth adding this for the sake of completeness

chrisfilo commented:
> @chrisfilo what keyword would you suggest be used instead of `acq`?

A new one as we discussed before. For example `megrel` (as in relation to MEG acquisition - `pre` or `post`). I am not hell bent on this though.

My main point is `pre` and `post` values need to have specific definitions and data curators should be required to use those values only and only if the cases described in the definitions apply. If we only suggest. If you really want to use `acq` that is ok, but in such case the following will apply:

a) For non `_markers` files curators can use any value of `acq` for any type of acqusiton
b) For `_markers` files curators can you any value of `acq` for any type of acqusiton as long as the values are not `pre` or `post`. `pre` or `post` can only be used to denote data acquisition before or after an MEG recording

Alternatively we can abandon the attempt to standardize (make it machine accessible) this aspect of metadata and just say that if you have two `_markers` files: one before and one after MEG recording you can use `acq` filename key with wahtever values you prefer to distinguish them.

yarikoptic commented:
How is it important to know the meaning/semantic for pre/post? 
 from #45 summary it seems that pre/post is just for cases when indeed multiple digitizations were done (so not even always), and then they likely just get averaged (so no semantic actually needed). So it sounds like too much battle for no winner ;-)

jasmainak commented:
yes, exactly my question too. In MNE, they are averaged anyway when reading the files. There are some cases according to @monkeyman192 when they can be analyzed separately but I would say it's the 20% rather than 80% use case. And at this point, we are probably bikeshedding :) 

For mne-bids, it's probably easier to allow acq to be free form. Otherwise, the user has to specify the pre and post labels manually. And the more work you give people for converting datasets, the less likely they will be to adapt the standard ...

monkeyman192 commented:
> Otherwise, the user has to specify the pre and post labels manually.

Not really. See https://github.com/mne-tools/mne-bids/pull/173

robertoostenveld commented:
Re "megrel" as name. The repeated acquisition of geometrical data (as the markers for the KIT system in this example) also happens in other contexts. E.g. repeated CTs to localize iEEG electrodes, repeated electrode digitizations using a Polhemus, repeated photo's for photogrammetry. Since this is all related to secondary data (the main data type is anat/meg/eeg/ieeg) it does not pertain to "runs" (as keyword). But it is also not MEG specific.

Re "What other situations do you envisage?" For example repeated measurements of the markers prior to an experiment (so pre1, pre2, pre3). Or measurements between runs (in case the subject comes out of the MEG system and/or coils that need to be reattached). 

I hope that we can find a nomenclature that can scale further than just "KIT markers", but more broadly meg and other modalities. We already have

    sub-0001_ses-01_acq-photo1_photo.jpg
    sub-0001_ses-01_acq-photo2_photo.jpg

in https://github.com/bids-standard/bids-specification/blob/master/src/04-modality-specific-files/04-intracranial-electroencephalography.md for me it feels natural to use that here as well. Please search for `acq-` on that page to see the different use cases.

chrisfilo commented:
I think I understand the context better now. Thank you.

You don't want to standardize how post/pre MEG marker files should be named. You just want to reiterate in the spec that `acq` can be used for avoiding name collisions for all sort of acquisition differences including relation to MEG. I think that is perfectly fine.

In such a case, I would rephrase the line https://github.com/bids-standard/bids-specification/pull/62/files#diff-56231ce79435a3bc626ef265c47d4acdR190 to include more broad language and treat `post`, `pre` as examples rather than recommendations.

monkeyman192 commented:
> In such a case, I would rephrase the line https://github.com/bids-standard/bids-specification/pull/62/files#diff-56231ce79435a3bc626ef265c47d4acdR190 to include more broad language and treat post, pre as examples rather than recommendations.

I'll try reword it to be more broad, but I still think that `pre` and `post` should be the recommended values.
But a recommendation is just that. People don't *need* to abide by it, but it would be nicer if they did.

chrisfilo commented on src/99-appendices/01-contributors.md:
```diff
@@ -100,6 +100,7 @@ your name is not listed, please add it.
 -   Jean-Baptiste Poline 📖📢🤔🎨
 -   Vasudev Raguram 💻🎨📖🔧
 -   Ariel Rokem 📖
+-   Matt Sanderson 📖💻
```
Please also add your github username to https://github.com/bids-standard/bids-specification/blob/master/CODEOWNERS

yarikoptic commented on src/99-appendices/06-meg-file-formats.md:
````diff
@@ -180,10 +180,19 @@ Where:
 ```Text
 sub-control01_ses-001_task-rest_run-01_meg/
-    sub-control01_ses-001_task-rest_run-01_markers.<mrk,sqd>
+    sub-control01_ses-001_task-rest_run-01[_acq-<acq-label>]_markers.<mrk,sqd>
````
here and elsewhere, should be `_acq-<label>` as per my unification through the entire specification a little back in #93

robertoostenveld commented:
Especially in the KIT case, to which this initially relates



> Op 28 feb. 2019 om 23:14 heeft monkeyman192 <notifications@github.com> het volgende geschreven:
> 
> In such a case, I would rephrase the line https://github.com/bids-standard/bids-specification/pull/62/files#diff-56231ce79435a3bc626ef265c47d4acdR190 to include more broad language and treat post, pre as examples rather than recommendations.
> 
> I'll try reword it to be more broad, but I still think that pre and post should be the recommended values.
> But a recommendation is just that. People don't need to abide by it, but it would be nicer if they did.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

chrisfilo commented:
+1

On Fri, Mar 1, 2019, 5:55 AM Robert Oostenveld <notifications@github.com>
wrote:

> Especially in the KIT case, to which this initially relates
>
>
>
> > Op 28 feb. 2019 om 23:14 heeft monkeyman192 <notifications@github.com>
> het volgende geschreven:
> >
> > In such a case, I would rephrase the line
> https://github.com/bids-standard/bids-specification/pull/62/files#diff-56231ce79435a3bc626ef265c47d4acdR190
> to include more broad language and treat post, pre as examples rather than
> recommendations.
> >
> > I'll try reword it to be more broad, but I still think that pre and post
> should be the recommended values.
> > But a recommendation is just that. People don't need to abide by it, but
> it would be nicer if they did.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bids-standard/bids-specification/pull/62#issuecomment-468672116>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOkpywTW5cKu6hV8fM4Ow-9OxBIWEA2ks5vSTE3gaJpZM4X0YyG>
> .
>

robertoostenveld commented on src/99-appendices/06-meg-file-formats.md:
````diff
@@ -180,10 +180,18 @@ Where:
     sub-control01_ses-001_task-rest_run-01_meg.<con,sqd>
 ```
 
+Marker coil files MUST have an acquisition parameter if there is more than
````
I propose to write it as 

If there are multiple files with marker coils, they must have the `acq-<label>` parameter and no more that two files may be associated with one raw data file. While the acquisition parameter can take any value, it is RECOMMENDED that if the two marker measurements occur before and after the raw data acquisition,
 `pre` and `post` are used to differentiate the two situations.

sappelhoff commented on src/99-appendices/06-meg-file-formats.md:
```diff
@@ -156,11 +156,11 @@ More about the 4D neuroimaging/BTi data organization at:
+contains coil positions in the acquisition system’s native space. Head points
+and marker points in head space are acquired using third-party hardware. One
+SHOULD rename/create a further run specific directory and keep the original
+files for each run inside.
```
this is colliding with #19 to some extend. Could you review that PR @monkeyman192?

monkeyman192 replied:
I'll wait for that PR to go through then rebase this onto the master to pick up the changes with the suggested rewording by @robertoostenveld

monkeyman192 commented:
@sappelhoff ok, rebased and hopefully addressed your comment as well as addressing @yarikoptic 's comment and @robertoostenveld 's comment (I did reword it a bit however as I think the wording wasn't quite right. Hopefully you agree!!)

sappelhoff commented on src/99-appendices/06-meg-file-formats.md:
```diff
@@ -173,10 +173,17 @@ sub-control01/
             sub-control01_ses-001_task-rest_run-01_meg.json
             sub-control01_ses-001_task-rest_run-01_channels.tsv
-            sub-control01_ses-001_task-rest_run-01_markers.<mrk,sqd>
+            sub-control01_ses-001_task-rest_run-01[_acq-<label>]_markers.<mrk,sqd>
```
Actually, according to the text in the [entities table appendix](https://bids-specification.readthedocs.io/en/stable/99-appendices/04-entity-table.html), the `acq` label must come before the `run` label :thinking: 

Could you change that @monkeyman192 ? (Sorry but I found it just now ...)

And it's another case where the validator is a bit too permissive, because when I tested it, it passed

sappelhoff replied:
```suggestion
            sub-control01_ses-001_task-rest[_acq-<label>]_run-01_markers.<mrk,sqd>
```

monkeyman192 replied:
yep, sorry, been sidetracked! Will get onto it asap!

effigies commented:
@jasmainak 
>  It would be nice if the CIs also build the doc and show it in the artifacts tab

https://496-150465237-gh.circle-artifacts.com/0/home/circleci/project/site/99-appendices/04-entity-table.html

jasmainak commented:
Oh nice, sorry missed it. Thanks! And PR looks good to me.

sappelhoff commented:
thanks @monkeyman192
